Implement reload config command

ericonr opened the pull request:
I rebased #52, solved conflicts, fixed all the issues that appeared, and addressed all of @emersion's pending comments in the original PR.

I'm not sure how I should group the commits together here, so would appreciate suggestions. Since this is the initial version, I left all my changes (besides the one mentioning varlink is optional in README) as additional commits; this way reviewers can hopefully have a clearer idea of what I changed. I can try and squash most of them together for a final version before pushing.

emersion commented on include/ipc.h:
```diff
@@ -0,0 +1,28 @@
+#include <stdlib.h>
+#include <varlink.h>
+
+#define varlink_return_if_fail_v(_val, _fmt, ...) { \
```
Let's not use macros

emersion commented on main.c:
```diff
@@ -16,8 +19,15 @@
+
 #define HEADS_MAX 64
 
+#define ARRAY_LENGTH(_arr) (sizeof(_arr) / sizeof(*(_arr)))
```
I'd prefer to just inline the `sizeof(arr) / sizeof(arr[0])` expressions

ericonr replied:
Done for event-loop.c

emersion commented on main.c:
```diff
@@ -479,6 +498,158 @@ static struct kanshi_config *read_config(const char *config) {
+}
+
+enum readfds_type {
+	FD_WAYLAND		= 0,
```
Style: don't align `=` signs. We can just omit the values completely here.

emersion commented on main.c:
```diff
@@ -479,6 +498,158 @@ static struct kanshi_config *read_config(const char *config) {
+	writefds[0].events = POLLOUT;
+
+	while (state->running) {
+	 while (wl_display_prepare_read(state->display) != 0) {
```
Style: indentation seems busted here

emersion commented on main.c:
```diff
@@ -479,6 +498,158 @@ static struct kanshi_config *read_config(const char *config) {
+};
+
+static int loop(struct kanshi_state *state) {
+	struct pollfd readfds[FD_COUNT];
```
Zero-initialize this array to avoid potential future pain

emersion commented on main.c:
```diff
@@ -479,6 +498,158 @@ static struct kanshi_config *read_config(const char *config) {
+	readfds[FD_VARLINK].events = POLLIN;
+#endif
+
+	struct pollfd writefds[1];
```
I wouldn't bother with `POLLOUT`

emersion commented on main.c:
```diff
@@ -479,6 +498,158 @@ static struct kanshi_config *read_config(const char *config) {
+	return false;
+}
+
+static int do_poll(struct pollfd *fds, nfds_t nfds) {
```
With `POLLOUT` gone, this can be inlined

emersion commented on main.c:
```diff
@@ -479,6 +498,158 @@ static struct kanshi_config *read_config(const char *config) {
+	FD_COUNT
+};
+
+static int loop(struct kanshi_state *state) {
```
I'd prefer to return `bool` values to indicate success/failure

ericonr replied:
This makes the logic slightly more complicated in `main`, which does `ret = kanshi_main_loop`. Not sure it'd be worth it?

emersion replied:
Fair enough

emersion commented on main.c:
```diff
@@ -507,21 +678,30 @@ int main(int argc, char *argv[]) {
 		}
 	}
 
+	struct wl_display *display = NULL;
```
This can be declared below, when first used.

emersion commented on main.c:
```diff
@@ -507,21 +678,30 @@ int main(int argc, char *argv[]) {
 		.running = true,
-		.config = config,
+		.display = display,
+		.config = config
```
Style nit: missing comma for last item

emersion commented on main.c:
```diff
@@ -532,12 +712,21 @@ int main(int argc, char *argv[]) {
-	while (state.running && wl_display_dispatch(display) != -1) {
-		// This space intentionally left blank
+	int ret = loop(&state);
+	goto done;
```
This `goto` spaghetti is a little confusing. Maybe remove the `error` label, and assign `ret = EXIT_FAILURE` at each error site?

ericonr replied:
As is it looks pretty ok I think?

emersion replied:
The gotos are interleaved which makes the logic tricky to follow.

emersion replied:
This appears to be fixed now.

emersion commented on parser.c:
```diff
@@ -33,6 +33,7 @@ static int parser_read_char(struct kanshi_parser *parser) {
 		return ch;
 	}
 
+	errno = 0;
```
This is unnecessary, `errno` is only checked when `fgetc` errors out, in which case it's guaranteed to be set to `fgetc's` error.

emersion commented:
Thanks for working on this!

> I'm not sure how I should group the commits together here, so would appreciate suggestions. 

Hm, I've made a bunch of review comments, but I only now realized you have may already fixed some of these in future commits. Sorry about that, feel free to disregard outdated comments.

Squashing commits together would help with the review process, because I don't want to review the full changes all together, I prefer to do one bit at a time. In an ideal world I'd split the commits into (1) add event loop (2) add signal handlers (3) add varlink interface (4) add kanshictl utility. But this may be too much work, I'm fine with a split into less commits. I'd still prefer to have logical commits rather than incremental commits (see https://www.bitsnbites.eu/git-history-work-log-vs-recipe/).

emersion commented:
Also, please retain attribution for the commits, but feel free to add `Co-authored-by` trailers to indicate when a commit has been worked on by multiple contributors.

Commit:
Factor out event loop.

This rework is necessary because it will allow us to add signal handlers
via signalfd and a varlink interface for IPC.

Co-authored-by: Érico Nogueira <erico.erc@gmail.com>

Commit:
Add signal handlers to reload the config.

Signaling kanshi with SIGHUP will now reload the config.

Co-authored-by: Érico Nogueira <erico.erc@gmail.com>

Commit:
Add optional varlink interface.

We don't want to fallback to libwayland's defaults, so WAYLAND_DISPLAY
needs to be set in env. Since IPC in the future is going to use
XDG_RUNTIME_DIR, we also require it.

Because these functions will be used in other parts of the code as well,
we rework the build system slightly to accommodate it.

Co-authored-by: Érico Nogueira <erico.erc@gmail.com>

Commit:
Add kanshictl utility.

It can talk to kanshi via the varlink interface. The only command
currently supported is to reload the config.

Co-authored-by: Érico Nogueira <erico.erc@gmail.com>

ericonr commented:
Luckily I had to do a similar refactor recently, so my git muscles were active already :)

It's done!

emersion commented on event-loop.c:
```diff
@@ -0,0 +1,65 @@
+			return EXIT_FAILURE;
+		}
+
+		if (wl_display_dispatch_pending(state->display) == -1) {
```
I think the `wl_display_read_events` + `wl_display_dispatch_pending` combo is the same as a single `wl_display_dispatch` call?

emersion replied:
That's incorrect. `wl_display_dispatch` also includes `wl_display_prepare_read_queue`.

Bleh. Why is this so complicated.

emersion commented on event-loop.c:
```diff
@@ -0,0 +1,65 @@
+			}
+		}
+
+		if (ret < 0 && errno != EPIPE) {
```
`EPIPE` should make us exit. Why is it special-cased?

emersion replied:
This code comes from `wl_display_dispatch_queue` in libwayland. Better to leave this as-is.

emersion commented on event-loop.c:
```diff
@@ -10,15 +10,77 @@
+
+static void signal_handler(int signum) {
+	if (write(signal_pipefds[1], &signum, sizeof(signum)) == -1) {
+		// we can't use stdio inside a signal handler
```
TBH, we can just call `abort()` here and stop trying to do anything more.

ericonr replied:
great, that simplifies things a bit :)

emersion commented on event-loop.c:
```diff
@@ -10,15 +10,77 @@
+static int set_pipe_flags(int fd) {
+	int flags = fcntl(fd, F_GETFL);
+	if (flags == -1) {
+		fprintf(stderr, "fnctl F_GETFL failed: %s\n", strerror(errno));
```
This can be simplified with `perror`. Same for all of the error messages below.

ericonr replied:
Oof, I completely missed that. Thanks

emersion commented on event-loop.c:
```diff
@@ -52,6 +114,36 @@ int kanshi_main_loop(struct kanshi_state *state) {
+					kanshi_reload_config(state);
+					break;
+				default:
+					return signum | 0x80;
```
Why this magic number?

ericonr replied:
This is from the original PR. I assume it came about from wanting the return value to show the program was killed by a signal. Will confirm it works portably and add a comment, otherwise just remove, I guess?

ericonr replied:
For now, I will consider it a success instead and just exit with 0 on signals...

emersion replied:
I'd be fine with that.

emersion commented on ipc-addr.c:
```diff
@@ -0,0 +1,20 @@
+#include "ipc.h"
+
+bool check_env(void) {
+	char *wayland_socket = getenv("WAYLAND_DISPLAY");
```
Nit: these can be `const`

emersion commented on meson.build:
```diff
@@ -43,18 +44,27 @@ kanshi_deps = [
 
+if varlink.found()
+	kanshi_deps += varlink
+	kanshi_args += '-DKANSHI_HAS_VARLINK'
```
I'd prefer to use `add_project_arguments` instead of `kanshi_args` to ensure the defines are propagated to the whole project, even if we change the build system in the future.

I'd also prefer to use `-DKANSHI_HAS_VARLINK=1` to indicate that libvarlink is available, and `DKANSHI_HAS_VARLINK=0` otherwise, then test with `#if` instead of `#ifdef`. Together with `-Wundef` this makes sure the define isn't lost.

ericonr replied:
Heh, I had exactly that issue when I was testing the PR :p 

One file had struct with varlink members, the other didn't. Using global args sounds much better indeed.

emersion commented on ipc-addr.c:
```diff
@@ -18,3 +18,8 @@ bool check_env(void) {
 	return true;
 }
+
+int get_ipc_address(char *address, size_t size) {
```
Maybe the `check_env` checks be moved to this function? We wouldn't rely on `check_env` being called prior to this one. Calling `get_ipc_address` without `check_env` results in a wrong address at best, or undefined behavior in the worst case (`printf` with NULL).

ericonr replied:
Sounds like a nice idea.

ericonr replied:
Hm, I did this, but it would mean we don't check `WAYLAND_DISPLAY` when IPC isn't built.

emersion replied:
Right. This is probably not that of a big deal.

emersion commented on meson_options.txt:
```diff
@@ -1 +1,2 @@
 option('man-pages', type: 'feature', value: 'auto', description: 'Generate and install man pages')
+option('varlink', type: 'feature', value: 'auto', description: 'Enable remote control with varlink')
```
Maybe rename this option to `ipc`? Whether varlink is used under the hood shouldn't really matter to the users.

emersion commented on meson.build:
```diff
@@ -81,6 +95,7 @@ if scdoc.found()
 	man_files = [
 		'kanshi.1.scd',
 		'kanshi.5.scd',
+		'kanshictl.1.scd',
```
Maybe only build this man page if ipc is enabled?

emersion commented on meson.build:
```diff
@@ -68,6 +68,20 @@ executable(
+			dependencies: [varlink],
+			c_args: kanshi_args,
+			install: true,
+		)
```
Indentation seems off here

emersion commented:
Overall this looks pretty good!

emersion commented on meson.build:
```diff
@@ -35,18 +35,42 @@ add_project_arguments(cc.get_supported_arguments([
+
+add_project_arguments([
+	'-DKANSHI_VERSION="@0@"'.format(meson.project_version()),
+	'-DKANSHI_HAS_VARLINK=@0@'.format(has_varlink),
```
This can be simplified to `varlink.found().to_int()`

emersion commented on event-loop.c:
```diff
@@ -10,6 +10,10 @@
 
 #include "kanshi.h"
 
+#ifdef KANSHI_HAS_VARLINK
```
All of these need to be `#if` instead of `#ifdef`, because `-DKANSHI_HAS_VARLINK=0` passes the `#ifdef` check.

Would be nice to add a new CI step with `--auto-features=disabled` passed to Meson, so that we can make sure the varlink-less build works fine (see wlroots CI which does this).

emersion commented on ipc-addr.c:
```diff
@@ -0,0 +1,21 @@
+#include "ipc.h"
+
+int get_ipc_address(char *address, size_t size) {
+	const char *wayland_socket = getenv("WAYLAND_DISPLAY");
```
Nit: can we rename this to `wayland_display`? `WAYLAND_SOCKET` is something else entirely, it's an already-opened FD picked up by libwayland.

ericonr committed:
Add CI step to test build without varlink.

ericonr commented:
I think it should be ready now, then :D

emersion reviewed (approved):
Sweet, thanks!
